### ISSUE TITLE
Pruning a sparse model

### ISSUE THREAD

**Issue by eldarkurtic** [neuralmagic/sparseml#145]
**Is your feature request related to a problem? Please describe.**
As far as I understand, pruning a sparse model is not supported. For example, a common use-case is to load a pruned model, with (for example) 70% sparsity, and to prune it additionally for 20% with (for example) `GMPruningModifier`. Currently, there is no way to initialize module masks (within the `GMPruningModifier`) from this sparse checkpoint.  

**Describe the solution you'd like**
I think that having a possibility to specify in your recipe that you want to load sparsity from the checkpoint would be really useful. For example:
```yaml
  - !GMPruningModifier
    params: __ALL__
    init_sparsity: 0.05
    final_sparsity: 0.8
    start_epoch: 0.0
    end_epoch: 36.0
    update_frequency: 1.0
    load_sparsity: True
```

**Describe alternatives you've considered**
1. alternative: Currently, it's possible to define a `ConstantPruningModifier` that would hold loaded module sparsities. This is beneficial only when additional finetuning of a sparse model is needed. Sparsity information contained in the `ConstantPruningModifier` is not accessible in other pruners (for example `GMPruningModifier`) and can't be used to prune an already sparse model. 
2. alternative: One could try to bypass this issue when initializing module masks by doing something like this:
```python
self._param_mask[self._param == 0.0] = 0.0
```
in the `__init__` of `ModuleParamPruningMask`. Unfortunately, one has to modify the `sparseml` source code to achieve this and to add the explicit trigger for this in his code, for example  command-line flag `--initialize-masks-from-ckpt`.

Solving this issue in the YAML recipe has the additional benefit for users to easily reproduce their experiments by having all necessary information stored in the recipe.

**Comment by markurtz** [comment]
@eldarkurtic, we added in the ability to load from a state_dict a few months ago as seen [here](https://github.com/neuralmagic/sparseml/blob/main/src/sparseml/pytorch/optim/modifier_pruning.py#L156). We currently don't have it added to the example scripts, though, so we can add that to the TODO list to enable.

Another proposal, which may get us most of the way there, is to support a flag for the init_sparsity such as FROM_PARAM, so on loading, it will start from whatever sparse structure that param is currently at and continue as normal.

**Comment by eldarkurtic** [comment]
Would the current load from a state_dict work if the checkpoint we are loading from is not previously pruned with `sparseml`? For example, there is a sparse model that is pruned in a different environment and we would like to prune it more with `sparseml`. Most likely it won't have any masks saved, but just raw weights (zero and non-zero weights).

And I agree that `init_sparsity: FROM_PARAM` would be the best solution for this, because loading a sparse model could cause some inconsistencies between the actual model and  the pruner's state (and there is no need for additional config params in the YAML recipe).

**Comment by markurtz** [comment]
@eldarkurtic we've added support in for FROM_PARAM now, closing this out. Thanks for the suggestion!
https://github.com/neuralmagic/sparseml/search?q=FROM_PARAM_TOKEN
